### DISCUSSION TITLE
Version 2 PySketcher API

### DISCUSSION THREAD

**Discussion by rvodden** [rvodden/pysketcher#286] [Announcements]
The current API requires you to know exactly where you'd like each of your shapes to appear. I think there is room for a more flexible API that describes the properties of the various shapes and PySketcher will then be able to calculate where these shapes could appear.

Right now I'm not exactly sure how that should look, so this discussion is a place to consider various options. 

For example we could use a fluent API like this:

```
p = Point(0,0)
c = Circle().centredAt(p).withRadius(1)
l1 = Line().horizontal().tangentTo(c)
l2 = Line().parallelTo(l2)
```

Another option would be to separate objects and constraints:

```
p = Point(0,0)
c = Circle()
l1 = Line()
l2 = Line()

con1 = CentredAtConstraint(c,p)
con2 = TangentialContraint(l1,c)
con3 = ParallelConstraint(l2,l1)
```

And then perhaps a combination of the two might be the correct answer:

```
p = Point(0,0)
c = Circle().centredAt(p).withRadius(5)
l1 = Line().tangentTo(c)
l2= Line().parallelTo(l2)
```

**Comment by rvodden** [comment]
Let's constrain the conversation a bit by focussing on Lines and Points to start with. I wonder how constructing a rectangle might best look.

```
height = 3
width = 5

p = Point(0,0)
l1 = Line().startsFrom(p).hasLength(height).isVertical()
l2 = Line().startsFrom(l1.end).hasLength(width).isHorizontal()
l3 = Line().startsFrom(l2.end).hadLength(l1.length).isParallelTo(l1)
l4 = Line().startsFrom(l3.ent).endsAt(l1.start)
```

My concern with this approach is that this puts a lot of onus on the Line object, so I wonder if there's a way of providing the Line object with helper classes that can implement the constraints.

**Comment by rvodden** [comment]
I've been working on a series of articles that explore this challenge in a little more depth:

[Part 1: Modelling Constraints](https://dev.to/rvodden/constraining-the-point-4pif)
[Part 2: More Modelling](https://dev.to/rvodden/writing-a-geometric-solver-in-python-part-2-more-modelling-44md)
[Part 3: Fixing our line](https://dev.to/rvodden/writing-a-geometric-solver-in-python-part-3-fixing-our-line-epk)

**Reply by ReblochonMasque** [reply]
This is pure genius!
I've read it, and then re-read it, and I think it is terrific.

**Reply by rvodden** [reply]
Hopefully I'll get part 4 out this weekend :)
